### ISSUE TITLE
MAINT: post __array_ufunc__ merger to-do list

### ISSUE THREAD

**Issue by mhvk** [numpy/numpy#8892]
EDITED to update with items from @eric-wieser and @shoyer. Feel free to edit to add other items.

A reminder of clean-up after #8247 that could most usefully be done:
- [x] Currently, arguments in ufuncs are normalized in `get_ufunc_arguments` (in `ufunc_object.c`), well after a check for an override is done; the latter code separately normalized arguments in `override.c`. Obviously, it would be better if only one of those was done (to the extend possible). As part of this, the type of exception raised should also be made consistent with python, i.e., `TypeError` for too many/too few arguments, things occurring both as positional and keyword arguments, etc.)

- [x] BUG: `__array_ufunc__` on arguments like `axis` would also override the ufunc (#9107).

- [x] `np.dot` and `np.matmul` are no longer overridden, but it would be good to override `np.matmul` in particular (as it implements a binop). This would need (1) turn `np.matmul` into a generalized ufunc (or have it use a new gufunc, after prepending/appending of axes as appropriate); (2) use the same generalized ufunc where feasible inside `np.dot` (and `ndarray.dot`), so that the most common cases can be overridden (for other cases, we might want to error on the presence of `__array_ufunc__`).

- [X] (@eric-wieser) Set the `__name__` field of the generated `NDArrayOperatorsMixin` methods (#9012)

- [X] (@shoyer) Need to switch ufunc overrides to always raise `TypeError` if any arguments set `__array_ufunc__ = None` (#9011)

- [X] (@shoyer) Choose the recommended way of implementing binary ops (calling the ufunc directly, or calling `__array_ufunc__`), and clarify guidance in the NEP (#9027)

- [ ] (@shoyer) Backwards compatible mixin like NDArrayOperatorsMixin as a separate project (#9016).

- [x] (@shoyer) BUG: no way to override matmul/@ if `__array_ufunc__` is set (#9028)

- [ ] (@mhvk) BUG: no way to override `__eq__` for structured `ndarray` (#9153)

- [ ] ENH/MAINT: let `ndarray.__array_ufunc__` be the overall gateway to applying ufuncs on `ndarray`; that way, this function can just participate in the different trials of objects to see if they can handle the ufunc. As part of that, should they just turn subclasses into regular arrays (since those have had a chance anyway? see discussion in #9079)

- [ ] (@shoyer) Document the table of ndarray arithmetic special methods and corresponding ufuncs in the proper docs (e.g., `reference/array.classes.rst`, not the NEP).

**Comment by eric-wieser** [comment]
Another one:

* [ ] Set the `__name__` field of the generated mixin class methods

**Comment by shoyer** [comment]
- [ ] Need to switch ufunc overrides to always raise `TypeError` if any arguments set `__array_ufunc__ = None` (https://github.com/numpy/numpy/issues/9011)
- [ ] Choose the recommended way of implementing binary ops (calling the ufunc directly, or calling `__array_ufunc__`), and remove the other recommendation from the NEP.

**Comment by mhvk** [comment]
@eric-wieser, @shoyer - I simply added your items to the list -- and deleted your comments to keep things a bit clearer. Will delete this one too...

**Comment by shoyer** [comment]
Backwards compatible mixin like NDArrayOperatorsMixin as a separate project (#9016).

**Comment by mhvk** [comment]
@mattip - the list is fairly complete, and all should at some point be solved...  But I think the `__matmul__` one is by far the most annoying/urgent..

**Comment by mattip** [comment]
updated - matmul is now a ufunc

**Comment by seberg** [comment]
I am not sure about this, is it considered a bug that `__array_ufunc__` currently will pass on additional keywords (not supported by numpy) happily, or is that intentional?

**Comment by mhvk** [comment]
I'd think it is a bug and (too implicitly) part of the first item above.

**Comment by seberg** [comment]
Looking at faster argparsing, I was starting to clean that up a bit (hopeully actually making it cleaner). That would "fix" that, but was not sure if I would have to add a cludge for it. Unfortunately it will be one hell of a merge conflict with my dtypes branch, but overall maybe it will make things easier anyway, since that needs a lot of cleanup as well.

**Comment by mhvk** [comment]
This is something I'm still interested in :smile: - possibly it can be made easier by aiming at the new fast call interface in python 3.8 (still private, but good for experiments; https://docs.python.org/3/whatsnew/3.8.html#vectorcall-a-fast-calling-protocol-for-cpython

p.s. My abandoned speed-up attempts are in #11372. Possibly individual commits are still useful. It definitely was a big mistake to try to do many things at once, though...

**Comment by seberg** [comment]
@mhvk yes, I am doing that, although my code right now has `#ifdef` to support both. If python is much faster on its side, it may help, otherwise I am not so sure. kwargs are pretty rare for usual ufunc calls (but very common for `reduce`). So for usual ufunc calls, my hope is currently mainly deleting some code duplication.

Anyway, halfway there, the review/some cleanup will be a pain though.

**Comment by seberg** [comment]
@mhvk geee, I did not realize that bunch of code/changes from you were lurking around. Will have a look at it, since you seemed to have gotten something out here, probably there are some thing I can copy over easily.

**Comment by seberg** [comment]
@mhvk I realize that I need to clean it up a bit, but would you be willing to take some look at the ufunc side of trying to get `METH_FASTCALL` at some point? (and `vectorcall`, but I am not sure it is really all that much worth it for now. `ufunc.__call__` is often called without keywords I think, and it has a pretty high overhead in any case).
It is a pretty huge diff though, so I can understand if its too much :/

An initial diff is here (I need to clean it out, and maybe squash it). Some of the things I did:

1. Parse all arguments to `object` initially
2. Always construct `full_args` (tuples of input and output arguments)
3. Use `full_args` tuples inside the overrides, so that it does not need to do any parsing/dict manipulation (unless one of the inputs/outputs supports `__array_ufunc__`).

This also moves the argument parsing to the initial `ufunc_generic_call` basically, and what you did also, moves the split betwen gufuncs up to there. (I am not 100% positive I already wrote the shim for the old `PyUfunc_GenericFunction`, which is an annoying C-API, that nobody uses.

https://github.com/seberg/numpy/compare/2d5b87e9f4837b70c068e832bd347b6b454e3bde..24ecc53f4d7a704d3ad0473d04550ba4539dabe6

**Comment by mhvk** [comment]
@seberg - I had an initial look and very much like the simplification in the processing order! I see you circumvented the annoyance I had of some functions being in the API which really are not handy any more (but likely are never used...).

One note on keyword arguments: if they were parsed as quickly as positional, it would actually become useful to pass `out` to ufuncs, which is much more readable than positional.

**Comment by seberg** [comment]
Thanks for having a look. I will split out the things into 3 (or 4) PRs, it hangs a bit on us being OK to duplicate some of the work that C-Python also semi-privdes with argument parsing.

Using kwargs are still quite a bit slower, but the speed difference is much smaller than before. For module level functions though, nothing much changes: `__array_function__` just adds too much overhead and cannot leverage `METH_FASTCALL`.

**Comment by seberg** [comment]
Marked first two as done, since gh-15271 covers them.

**Comment by seberg** [comment]
Going to close this, I don't think keep this open serves much purpose anymore:
* Yes, we can add more documentation...
* The ABC stuff might be intersting (not sure), but also has its own issue.
* Overriding `__eq__` for voids, should not likely be solved by making void comparisons a ufunc (non-trivial, but I still think that is probably the right way)
* The idea of refactoring how `__array_ufunc__` entry-point may be interesting, but I am not sure and keeping it around here won't make it happen.

**Comment by mhvk** [comment]
Yes, makes sense to close -  the one that is still a buglet has its own issue anyway.
